lib/AnalysisStructured/PtrAnalysis.cpp

```diff
@@ -0,0 +1,1128 @@
+
+LogicalResult PtrSState::addState(const PtrSState &lhsState,
+                                 const PtrSState &rhsState, Operation *op,
+                                 OpBuilder &builder) {
+  assert(isEmpty() && lhsState.getRank() == rhsState.getRank());
+  auto loc = op->getLoc();
+
+  if (lhsState.source && rhsState.source) {
+    op->emitRemark(
+        "PtrAnalysis: do not support adding two pointer states that both "
+        "have base pointers");
+    return failure();
+  }
+
+  source = lhsState.source ? lhsState.source : rhsState.source;
+
+  // AddPtr where both lhs and rhs containing modulo operators not supported
+  if (lhsState.hasModulo() && rhsState.hasModulo()) {
+    op->emitRemark("PtrAnalysis: do not support adding two pointer states "
+                   "that both have modulo");
+    return failure();
+  }
+
+  if (lhsState.hasModulo() || rhsState.hasModulo()) {
+    // visitOperandSplat and visitOperandExpandDims should enforce below
+    assert(lhsState.getRank() <= 2);
+  }
+
+  if (lhsState.scalar && rhsState.scalar) {
+    auto addOp =
+        builder.create<arith::AddIOp>(loc, lhsState.scalar, rhsState.scalar);
+    scalar = addOp.getResult();
+  } else if (lhsState.getRank() == 0) { // both lhs and rhs are scalars
+    scalar = lhsState.scalar ? lhsState.scalar : rhsState.scalar;
+  }
+
+  for (uint64_t i = 0; i < lhsState.getRank(); i++) {
+    auto newOffset =
+        addOFRs(lhsState.offsets[i], rhsState.offsets[i], loc, builder);
+    offsets.push_back(newOffset);
+
+    auto newStride =
+        addOFRs(lhsState.strides[i], rhsState.strides[i], loc, builder);
+    strides.push_back(newStride);
+
+    sizes.push_back(lhsState.sizes[i]);
+  }
+
+  // dealing with modulo:
+  // - If lhs has no modulo, skip
+  // - If rhs has zero offset on dim i, we can just use lhs's modulo
+  // - If i == 0 and rhs is the result of a splat, we will allow the add. This
+  // is because the user may be trying to express adding a constant offset to
+  // increment dim1, but pointer analysis cannot differentiate dim1 vs dim0 in
+  // this case.
+  // - Else, the analysis fail
+
+  // Note that this is not bullet-proof. E.g., broken IR can actually increment
+  // dim0 while dim0 already has modulo, since Triton offsets are element-wise
+  // and not in unit of lower dimensions. However, this is highly unlikely but
+  // the analysis will provide wrong result. Hence we provide a warning in this
+  // case.
+  PtrSState const *lhs = &lhsState;
+  PtrSState const *rhs = &rhsState;
+
+  if (rhs->hasModulo()) {
+    std::swap(lhs, rhs);
+  }
+
+  for (uint64_t i = 0; i < lhs->getRank(); i++) {
+    if (!lhs->dimHasModulo(i)) {
+      modulos.push_back(lhs->modulos[i]);
+    } else if (hasConstZero(rhs->offsets[i])) {
+      modulos.push_back(lhs->modulos[i]);
+    } else if (i == 0 && lhs->getRank() == 2 && rhs->scalar) {
+      modulos.push_back(lhs->modulos[1]);
+      modulos.push_back(lhs->modulos[0]);
+      op->emitWarning(
```

Comment:
Should we add a strictly-correct option and fail this case? Afaik this is the only transformation in this pass that can lead to potentially incorrect result.